src/coreclr/vm/pendingload.cpp

```diff
@@ -12,194 +12,26 @@
 
 #ifndef DACCESS_COMPILE
 
+#ifdef PENDING_TYPE_LOAD_TABLE_STATS
 
-// ============================================================================
-// Pending type load hash table methods
-// ============================================================================
-/*static */ PendingTypeLoadTable* PendingTypeLoadTable::Create(LoaderHeap *pHeap,
-                                                               DWORD dwNumBuckets,
-                                                               AllocMemTracker *pamTracker)
+LONG pendingTypeLoadEntryDynamicAllocations = 0;
+void PendingTypeLoadEntryDynamicAlloc()
 {
-    CONTRACTL
-    {
-        STANDARD_VM_CHECK;
-        PRECONDITION(CheckPointer(pHeap));
-    }
-    CONTRACTL_END;
-
-    size_t                  size = sizeof(PendingTypeLoadTable);
-    BYTE *                  pMem;
-    PendingTypeLoadTable * pThis;
-
-    _ASSERT( dwNumBuckets >= 0 );
-    S_SIZE_T allocSize = S_SIZE_T( dwNumBuckets )
-                                        * S_SIZE_T( sizeof(PendingTypeLoadTable::TableEntry*) )
-                                        + S_SIZE_T( size );
-    if( allocSize.IsOverflow() )
-    {
-        ThrowHR(E_INVALIDARG);
-    }
-    pMem = (BYTE *) pamTracker->Track(pHeap->AllocMem( allocSize ));
-
-    pThis = (PendingTypeLoadTable *) pMem;
-
-#ifdef _DEBUG
-    pThis->m_dwDebugMemory = (DWORD)(size + dwNumBuckets*sizeof(PendingTypeLoadTable::TableEntry*));
-#endif
-
-    pThis->m_dwNumBuckets = dwNumBuckets;
-    pThis->m_pBuckets = (PendingTypeLoadTable::TableEntry**) (pMem + size);
-
-    // Don't need to memset() since this was ClrVirtualAlloc()'d memory
-    // memset(pThis->m_pBuckets, 0, dwNumBuckets*sizeof(PendingTypeLoadTable::TableEntry*));
-
-    return pThis;
+    InterlockedIncrement(&pendingTypeLoadEntryDynamicAllocations);
 }
+#endif // PENDING_TYPE_LOAD_TABLE_STATS
 
+static BYTE s_PendingTypeLoadTable[sizeof(PendingTypeLoadTable)];
```

Comment:
This is really an anti-pattern and can create UB easily. Looking at the type there shouldn't be any reason to not declare as:

```cpp
static PendingTypeLoadTable s_PendingTypeLoadTable;
```

Reply:
I'm pretty sure that `Crst` types can't be in static variables, and they are all over this thing. However, I will check before my next update.

Reply:
We have `CrstStatic` that is a destructor-free variant of Crst that is meant to be used for static variables.

Reply:
Ah, but most of the `Crst` instances are on the Entry structures. Those are structures which both are allocated statically, as well as dynamically in rare cases. I think that makes `CrstStatic` invalid for those scenarios. However, `CrstStatic` could be used for the variables in the PendingTypeLoadTable, and I'll look into if we can do something about the `Entry` structs.

Reply:
I was able to get it to work. A bit gross, but its fine.